Allow optional persistent datadir

rajarshimaitra opened the pull request:
The intention here is to open possibility to have persistent data directory option for the node along with tempdir option too..

There were many ways of doing this.. I choose this one as it seemed the one with least changes..

Looking for approach acks..

RCasatta commented:
Internals logic with `DataDir` structs looks ok from a first look but I have doubts with the `Conf` struct.

Use cases we need are:
1) Create a temp dir where the SO put temporary stuff (eg /tmp)
2) Create a temp dir in this root directory I am giving you
3) Use this datadir I am giving you and avoid deletion at the end

The reason why `Conf` is `#[non_exhaustive]` is to be free to add fields while avoiding breaking changes downstream, if we rename `tmpdir` in `datadir` not only we are not compatible, but we are loosing use case (2)

I think we should add a field in the `Conf` struct which is `datadir: Option<PathBuf>`

If both `datadir` and `tmpdir` are some we throw an error

if both are none we do (1)

if tmpdir is some we do (2)

if datadir is some we do (3)

rajarshimaitra commented:
Yes thanks for the suggestion. That makes perfect sense. Updated the `Conf` struct to simulate the behavior..

RCasatta commented on src/lib.rs:
```diff
@@ -168,6 +201,7 @@ impl Default for Conf<'_> {
             p2p: P2P::No,
             network: "regtest",
             tmpdir: None,
+            datadir: None,
```
There is a doctest for `Conf` struct where default values are documented, could you add a line `conf.datadir = None;`

rajarshimaitra replied:
Done..

RCasatta commented on src/lib.rs:
```diff
@@ -183,16 +217,23 @@ impl BitcoinD {
-        let cookie_file = datadir.join(conf.network).join(".cookie");
+        let work_dir = match (&conf.tmpdir, &conf.datadir) {
+            (Some(_), Some(_)) => {
+                return Err(Error::Custom(
```
Could you please create a specific error instead of a generic `Custom` variant?
(same text is ok in the `fmt::Debug` is ok)

RCasatta replied:
I try to avoid error variants with reason strings because they are harder to match against the specific reason downstream.
Please use a variant like `BothDirsSpecified` which formats to "tempdir and staticdir cannot be enabled at same time" as you wrote for `reason`

RCasatta commented on src/lib.rs:
```diff
@@ -145,13 +167,24 @@ pub struct Conf<'a> {
+    /// tmpdir is Some() && workdir is Some() : Error. Cannot be enabled at same time.
+    /// tmpdir is Some(temp_path) && workdir is None : Creates temporary directory at temp_path.
+    /// tmpdir is None && workdir is Some(work_path) : Creates persistent directory at work_path.
+    /// tmpdir is None && workdir is None: Creates temporary directory at `/tmp` or `TEMPDIR_ROOT` env variable path.
```
Creates a temporary directory in OS default temporary directory (eg `/tmp`)

RCasatta commented on src/lib.rs:
```diff
@@ -300,6 +341,11 @@ impl BitcoinD {
     }
 
+    /// Return the current workdir path of the running node
+    pub fn get_datadir_path(&self) -> PathBuf {
```
just `datadir()` ?

RCasatta reviewed (commented):
Could you please also squash the 2 commits because the second one reworks stuff of the previous?

Thanks for your work

RCasatta commented on src/lib.rs:
```diff
@@ -35,12 +35,31 @@ pub struct BitcoinD {
     pub client: Client,
     /// Work directory, where the node store blocks and other stuff. It is kept in the struct so that
```
needs update to account for the persisted case

rajarshimaitra replied:
Removed the last line all together. More info on `DataDir` is in its own doc and usage info is in `Conf` doc..

rajarshimaitra commented:
Thanks @RCasatta for the review.. All were valid points.. Polished up the code a bit.. Changed the name `workdir` to `staticdir`, similar names were getting confusing.. 

Added the `self.client.stop` at `Drop`, but not sure what happens when the client returns error in a stop call.. Maybe it just ignores and drops anyway..

RCasatta commented:
> Added the self.client.stop at Drop, but not sure what happens when the client returns error in a stop call.. Maybe it just ignores and drops anyway..

I think it's fine how you did it (no panic should ever happen in drop)

utack with https://github.com/RCasatta/bitcoind/pull/52#discussion_r832068948

rajarshimaitra commented:
> utack except for https://github.com/RCasatta/bitcoind/pull/52#discussion_r832068948

I assume its the comment on the specific error type? Fixed that in last push..

RCasatta commented:
ACK 7e05242d2b128ec7152b709277f01b495fdd5446

thanks!
